runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -56,181 +57,436 @@ extern "C" {
-        // are passed in GPRs 1-3. If we have passed those limits while iterating through the argument list, then we can stop
-        // counting the number of instructions. We won't need to store the remaining parameters because they are already
-        // in memory.
-        if (fprArgCount>=4 && gprArgCount>=3)
-            break;
+	// Loop through the arguments and count how many instructions you'll need to copy data from registers to memory, and how many
+	// bytes of space you'll need to allocate in thunk and new stack frame to store the instructions + parameter info.
+	for (I_32 i = 0; i < lastSigIdx; i++) {
+		U_32 argSizeInBytes = sigArray[i].sizeInByte;
+		switch (sigArray[i].type) {
+		case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_INT64:
+			if (gprArgCount < maxGPRArgCount) {
+				numInstructions++;
+				gprArgCount++;
+			}
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+		case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+			if (fprArgCount < maxFPRArgCount) {
+				numInstructions++;
+				fprArgCount++;
+			}
+			if (i < 3) {
+				gprArgCount++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+			}
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_ALL_SP:
+			if (argSizeInBytes == 2 * sizeof(float)) {
+				// a pure float struct with only 2 floats is passed in 2 FPRs.
+				// If only one FPR is free, then the first half of the struct is
+				// in FPR6, and the second half is in the corresponding slot
+				// in the argument area, right adjusted.
+				if (fprArgCount <= maxFPRArgCount - 2) {
+					numInstructions += 2;
+					fprArgCount += 2;
+				}
+				else if (fprArgCount < maxFPRArgCount) {
+					numInstructions++;
+					fprArgCount++;
+				}
+			}
+			else { // All other pure float structs are passed in GPR (if there's space)
+				if (gprArgCount < maxGPRArgCount) {
+					// gprsNeededForEntireStruct below represents the number of GPRs we would need hypothetically to load the entire struct in GPRs.
+					// Since this is a pure float structure, the size will always be a multiple of 4. So if we divide by 8, like below, the result
+					// will either be some whole number 'X', or 'X.5'. For `X.5`, adding 0.5 to the sum will always round UP to the nearest whole number.
+					int gprsNeededForEntireStruct = (int)(((float)argSizeInBytes / 8.0F) + 0.5);
 
-    }
+					// We may not have all the GPRs available to store the entire struct. In that case, we store only what's in GPR (the rest of the struct will already be in memory.)
+					int numInstructionsAvailable = maxGPRArgCount - gprArgCount;
```

Comment:
So I think this comes down to preference (as variable naming is in general).

Before opening this PR I thought about renaming it something similar to what you suggested. In the end I went with the other option.

The reason is that the calculation for this variable comes from GPR counters, so it's quite self explanatory that it's showing the number of available GPRs. What's not self explanatory, at least to someone looking at the code for the first time, is why we need the count of GPRs available (i.e. it also represents the number of instructions we can generate). That's why I went with naming it the way it is.

However, in the end it's just a naming change. I don't feel that strongly about naming it. Just wanted to explain my thinking around why I chose the naming I did. If you still feel strongly about naming it from a GPR perspective, then I'll change it. Please let me know. Thanks.

Reply:
`numInstructionsAvailable` suggests that we have limited instructions and the calculation is about remaining instructions we can generate. I think for readability - numOfAvailable GPRs makes sense, even though calculation is self explanatory, renaming it would allow better readable code where you have used that variable.

Reply:
Naming should be changed now in b6cac56c59ac275b1cd08ba52da968d2a7cb5e94.

Reply:
Forgot a few places in the above commit. I've addressed the remaining places here: 1d44f973174c463fe3a0655ce92207bea6fe6259